### PR TITLE
Implement RFC 1542

### PR DESCRIPTION
**Pull request by sfackler**
cc #33417 

r? @aturon

### CODE REVIEW

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -2341,9 +2341,101 @@ macro_rules! from_str_radix_int_impl {
...
+#[derive(Debug, Copy, Clone)]
+pub struct TryFromIntError(());
+
+impl TryFromIntError {
+    #[unstable(feature = "int_error_internals",
+               reason = "available through Error trait and this method should \
+                         not be exposed publicly",
+               issue = "0")]
+    #[doc(hidden)]
+    pub fn __description(&self) -> &str {
```
Comment (aturon) [inline src/libcore/num/mod.rs]: Yuck! But I guess such is life for private sharing across the facade...

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -2341,9 +2341,101 @@ macro_rules! from_str_radix_int_impl {
...
+
+#[unstable(feature = "try_from", issue = "33417")]
+impl fmt::Display for TryFromIntError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.__description().fmt(fmt)
+    }
+}
+
+macro_rules! same_sign_from_int_impl {
+    ($storage:ty, $target:ty, $($source:ty),*) => {$(
```
Comment (aturon) [inline src/libcore/num/mod.rs]: This macro syntax could probably be more clear (given the open-ended arg list, it's hard to see what's going on from a glance at the use-site). But NBD.

**Comment by aturon** [comment]
Basically LGTM, but I'd like tests for the edge cases on the numeric conversions.

**Comment by sfackler** [comment]
Yep, will add tests tonight

File: src/libcore/num/mod.rs (hunk 1 above)
Comment (sfackler) [inline src/libcore/num/mod.rs]: Yeah, this is what the existing `ParseIntError` did :(

**Comment by kamalmarhubi** [comment]
Thanks for doing this @sfackler, I'm quite excited for these traits. :-)

**Comment by sfackler** [comment]
Added some tests

@bors r=aturon

**Pull request merged**